File: app/routes/authRoutes.ts
// src/routes/authRoutes.ts
// Maps to: /auth/* endpoints in auth_routes.py

const BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

// ── Types ─────────────────────────────────────────────────────────────────────

export interface SignupPayload {
  // Required
  username: string;
  email: string;
  password: string;
  pharmacy_name: string;
  // Optional
  contact_number?: string | null;
  whatsapp_number?: string | null;
  address?: string | null;
  opening_hours?: string | null;
  estimated_delivery_time?: string | null;
  service_areas?: string | null;
  prescription_policy?: string | null;
  refund_policy?: string | null;
}

export interface LoginPayload {
  username_or_email: string; // accepts username OR email
  password: string;
}

export interface UserResponse {
  id: number;
  username: string;
  email: string;
  created_at: string;
}

export interface PharmacyResponse {
  id: number;
  user_id: number;
  pharmacy_name: string;
  contact_number: string | null;
  whatsapp_number: string | null;
  address: string | null;
  opening_hours: string | null;
  estimated_delivery_time: string | null;
  service_areas: string | null;
  prescription_policy: string | null;
  refund_policy: string | null;
  created_at: string;
}

export interface SignupResponse {
  message: string;
  user: UserResponse;
  pharmacy: PharmacyResponse;
}

export interface LoginResponse {
  message: string;
  access_token: string;
  token_type: "bearer";
  user: UserResponse;
  pharmacy: PharmacyResponse;
}

export interface MeResponse {
  user: UserResponse;
  pharmacy: PharmacyResponse;
  token_claims: {
    user_id: number;
    pharmacy_id: number;
    exp: number;
  };
}

// ── Helper ────────────────────────────────────────────────────────────────────

async function handleResponse<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: "Request failed" }));
    throw new Error(err.detail ?? "Request failed");
  }
  return res.json() as Promise<T>;
}

// ── POST /auth/signup ─────────────────────────────────────────────────────────
// Creates a new user + pharmacy profile in one call.
// Backend: auth_routes.py → signup()
// Only required fields: username, email, password, pharmacy_name.
// All other fields are optional and can be set later via PUT /pharmacy/me.
export async function signup(payload: SignupPayload): Promise<SignupResponse> {
  const res = await fetch(`${BASE}/auth/signup`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return handleResponse<SignupResponse>(res);
}

// ── POST /auth/login ──────────────────────────────────────────────────────────
// Authenticates a user. Returns a JWT access_token (custom HMAC, not JWT lib).
// Token must be saved and sent as: Authorization: Bearer <token>
// Token expires in 24 hours (86400s) — see auth_token_service.py.
// Backend: auth_routes.py → login()
export async function login(payload: LoginPayload): Promise<LoginResponse> {
  const res = await fetch(`${BASE}/auth/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return handleResponse<LoginResponse>(res);
}

// ── GET /auth/me ──────────────────────────────────────────────────────────────
// Returns the currently authenticated user + pharmacy profile.
// Requires a valid Bearer token in the Authorization header.
// Backend: auth_routes.py → auth_me()
export async function getMe(token: string): Promise<MeResponse> {
  const res = await fetch(`${BASE}/auth/me`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });
  return handleResponse<MeResponse>(res);
}

// ── Token helpers (localStorage) ──────────────────────────────────────────────
// Convenience functions to save/read/clear auth state in the browser.

export function saveAuthToStorage(data: LoginResponse): void {
  localStorage.setItem("healix_token",    data.access_token);
  localStorage.setItem("healix_user",     JSON.stringify(data.user));
  localStorage.setItem("healix_pharmacy", JSON.stringify(data.pharmacy));
}

export function getTokenFromStorage(): string | null {
  return localStorage.getItem("healix_token");
}

export function getUserFromStorage(): UserResponse | null {
  const raw = localStorage.getItem("healix_user");
  return raw ? (JSON.parse(raw) as UserResponse) : null;
}

export function getPharmacyFromStorage(): PharmacyResponse | null {
  const raw = localStorage.getItem("healix_pharmacy");
  return raw ? (JSON.parse(raw) as PharmacyResponse) : null;
}

export function clearAuthStorage(): void {
  localStorage.removeItem("healix_token");
  localStorage.removeItem("healix_user");
  localStorage.removeItem("healix_pharmacy");
}

export function isLoggedIn(): boolean {
  return !!getTokenFromStorage();
}

File: PythonProject(Stock mannagement)/app/routes/auth_routes.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer

from app.database.user_db import UserSessionLocal
from app.services.auth_service import AuthService
from app.services.auth_token_service import create_access_token, decode_access_token

router = APIRouter(prefix="/auth", tags=["Auth"])
security = HTTPBearer()


class SignupRequest(BaseModel):
    username: str
    email: str
    password: str
    pharmacy_name: str
    contact_number: str | None = None
    whatsapp_number: str | None = None
    address: str | None = None
    opening_hours: str | None = None
    estimated_delivery_time: str | None = None
    service_areas: str | None = None
    prescription_policy: str | None = None
    refund_policy: str | None = None


class LoginRequest(BaseModel):
    username_or_email: str
    password: str


def get_user_db():
    db = UserSessionLocal()
    try:
        yield db
    finally:
        db.close()


def _serialize_user(user):
    return {
        "id": user.id,
        "username": user.username,
        "email": user.email,
        "created_at": user.created_at
    }


def _serialize_pharmacy(pharmacy):
    return {
        "id": pharmacy.id,
        "user_id": pharmacy.user_id,
        "pharmacy_name": pharmacy.pharmacy_name,
        "contact_number": pharmacy.contact_number,
        "whatsapp_number": pharmacy.whatsapp_number,
        "address": pharmacy.address,
        "opening_hours": pharmacy.opening_hours,
        "estimated_delivery_time": pharmacy.estimated_delivery_time,
        "service_areas": pharmacy.service_areas,
        "prescription_policy": pharmacy.prescription_policy,
        "refund_policy": pharmacy.refund_policy,
        "created_at": pharmacy.created_at,
    }


def get_current_auth_context(
    credentials: HTTPAuthorizationCredentials = Depends(security),
    db: Session = Depends(get_user_db),
):
    try:
        payload = decode_access_token(credentials.credentials)
    except ValueError as exc:
        raise HTTPException(status_code=401, detail=str(exc))

    service = AuthService(db)
    user = service.get_user_by_id(int(payload["sub"]))
    if not user:
        raise HTTPException(status_code=401, detail="User not found")

    pharmacy = service.get_pharmacy_by_user_id(user.id)
    if not pharmacy:
        raise HTTPException(status_code=404, detail="Pharmacy profile not found")

    return {"user": user, "pharmacy": pharmacy, "payload": payload}


@router.post("/signup")
def signup(payload: SignupRequest, db: Session = Depends(get_user_db)):
    service = AuthService(db)
    try:
        user, pharmacy = service.signup(
            username=payload.username,
            email=payload.email,
            password=payload.password,
            pharmacy_name=payload.pharmacy_name,
            contact_number=payload.contact_number,
            whatsapp_number=payload.whatsapp_number,
            address=payload.address,
            opening_hours=payload.opening_hours,
            estimated_delivery_time=payload.estimated_delivery_time,
            service_areas=payload.service_areas,
            prescription_policy=payload.prescription_policy,
            refund_policy=payload.refund_policy,
        )
        return {
            "message": "Signup successful",
            "user": _serialize_user(user),
            "pharmacy": _serialize_pharmacy(pharmacy),
        }
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


@router.post("/login")
def login(payload: LoginRequest, db: Session = Depends(get_user_db)):
    service = AuthService(db)
    try:
        user = service.login(
            username_or_email=payload.username_or_email,
            password=payload.password
        )
        pharmacy = service.get_pharmacy_by_user_id(user.id)
        if not pharmacy:
            raise HTTPException(status_code=404, detail="Pharmacy profile not found")

        access_token = create_access_token(user_id=user.id, pharmacy_id=pharmacy.id)
        return {
            "message": "Login successful",
            "access_token": access_token,
            "token_type": "bearer",
            "user": _serialize_user(user),
            "pharmacy": _serialize_pharmacy(pharmacy),
        }
    except ValueError as exc:
        raise HTTPException(status_code=401, detail=str(exc))


@router.get("/me")
def auth_me(context=Depends(get_current_auth_context)):
    return {
        "user": _serialize_user(context["user"]),
        "pharmacy": _serialize_pharmacy(context["pharmacy"]),
        "token_claims": {
            "user_id": context["payload"]["sub"],
            "pharmacy_id": context["payload"]["pharmacy_id"],
            "exp": context["payload"]["exp"],
        }
    }
